storage/src/vespa/storage/persistence/filestorage/filestormanager.cpp

```diff
@@ -973,18 +981,63 @@ void FileStorManager::initialize_bucket_databases_from_provider() {
     _init_handler.notifyDoneInitializing();
 }
 
+class FileStorManager::TrackExecutedTasks : public vespalib::Executor::Task {
+public:
+    TrackExecutedTasks(FileStorManager & manager);
+    void run() override;
+private:
+    FileStorManager & _manager;
+    size_t            _serialNum;
+};
+
+FileStorManager::TrackExecutedTasks::TrackExecutedTasks(FileStorManager & manager)
+    : _manager(manager),
+      _serialNum(0)
+{
+    std::lock_guard guard(_manager._executeLock);
+    _serialNum = _manager._executeCount++;
+    _manager._tasksInExecute.insert(_serialNum);
+}
+
+void
+FileStorManager::TrackExecutedTasks::run() {
+    std::lock_guard guard(_manager._executeLock);
+    _manager._tasksInExecute.erase(_serialNum);
+}
+
 std::unique_ptr<spi::BucketTask>
 FileStorManager::execute(const spi::Bucket &bucket, std::unique_ptr<spi::BucketTask> task) {
     StorBucketDatabase::WrappedEntry entry(_component.getBucketDatabase(bucket.getBucketSpace()).get(
             bucket.getBucketId(), "FileStorManager::execute"));
     if (entry.exist()) {
-        _filestorHandler->schedule(std::make_shared<RunTaskCommand>(bucket, std::move(task)));
+        auto trackBuckets = std::make_unique<TrackExecutedTasks>(*this);
+        _filestorHandler->schedule(std::make_shared<RunTaskCommand>(bucket, std::move(trackBuckets), std::move(task)));
     }
     return task;
 }
 
+bool
+FileStorManager::areTaskCompleteUntil(size_t limit) const
+{
+    std::lock_guard guard(_executeLock);
+    for (size_t serial : _tasksInExecute) {
+        if (serial < limit) {
+            return false;
+        }
+    }
+    return true;
+}
+
 void
 FileStorManager::sync() {
+    size_t serialNumLimit;
+    {
+        std::lock_guard guard(_executeLock);
+        serialNumLimit = _executeCount;
+    }
+    while ( ! areTaskCompleteUntil(serialNumLimit)) {
+        std::this_thread::sleep_for(100us);
```

Comment:
Will `sync` be called often enough that it'd make sense to have some kind of conditional thread notify mechanism instead?

Reply:
It is currently called very rarely, but it might be better to use notify.

Reply:
Fixed